### PR TITLE
Panic rtt target riscv

### PR DESCRIPTION
**Pull request by perlindgren**
CS based `panic_rtt_target`.
- platform agnostic
- addressed comments
- updated example
- updated documentation
- avoid storing `SEGGER RTT` verbatim to avoid misinterpretation of RttHeader location

Fixed some minor issues in `rtt_target` to get it working.

### CODE REVIEW

File: panic-rtt-target/src/lib.rs [hunk 1]
```diff
@@ -50,29 +50,29 @@ use core::{
...
-    use cortex_m::interrupt;
+    critical_section::with(|_| {
+        if let Some(mut channel) = unsafe { UpChannel::conjure(0) } {
+            channel.set_mode(ChannelMode::BlockIfFull);
 
-    interrupt::disable();
+            writeln!(channel, "{}", info).ok();
+        } else {
+            // failed to get channel, but not much else we can do but spin
+            loop {
```
Comment (Dirbaio) [inline panic-rtt-target/src/lib.rs:63]: this inifinite loop seems redundant with the next one.

File: rtt-target/src/rtt.rs [hunk 2]
```diff
@@ -74,8 +66,9 @@ impl RttChannel {
         ptr::write_volatile(&mut self.buffer, buffer as *mut u8);
     }
 
+    /// True if buffer NOT pointing to null
     pub fn is_initialized(&self) -> bool {
-        self.buffer.is_null()
+        !self.buffer.is_null()
```
Comment (Dirbaio) [inline rtt-target/src/rtt.rs:71]: Huh, this is interesting. Could you send it as a separate PR?

File: panic-rtt-target/src/lib.rs [hunk 3]
```diff
@@ -50,29 +50,29 @@ use core::{
...
-        writeln!(channel, "{}", info).ok();
-    }
+        // we should never leave critical section
+        loop {
+            compiler_fence(SeqCst);
+        }
+    });
 
+    // will in fact be unreachable
     loop {
```
Comment (Dirbaio) [inline panic-rtt-target/src/lib.rs:75]: I think this one should be removable as well, if you keep the one inside

**Comment by Dirbaio** [comment]
Could you remove the unrelated changes to `rtt-target`?

File: rtt-target/src/rtt.rs [hunk 4]
```diff
@@ -29,18 +29,10 @@ impl RttHeader {
...
         ptr::write_volatile(&mut self.max_down_channels, max_down_channels);
 
-        // Copy the ID in two parts to avoid having the ID string in memory in full. The ID is
-        // copied last to make it less likely an unfinished control block is detected by the host.
-
-        ptr::copy_nonoverlapping(b"SEGG_" as *const u8, self.id.as_mut_ptr(), 5);
-
+        // The ID is copied last to make it less likely an unfinished control block is detected by the host.
+        let s = b"SEGGER RTT\0\0\0\0\0\0";
+        ptr::copy_nonoverlapping(s as *const u8, self.id.as_mut_ptr(), s.len());
```
Comment (mvirkkunen) [inline rtt-target/src/rtt.rs:34]: This write was done in two parts with the extra character overwritten on purpose to avoid ever having the magic string in memory, even as a temporary. Not sure if that's actually required or useful though, but I was being careful.

File: panic-rtt-target/src/lib.rs (hunk 1 above)
Comment (perlindgren) [reply panic-rtt-target/src/lib.rs:63]: Its interesting to see where it got stuck. (If there was a problem getting the channel.) Assembly wise it should be a single branch to itself (in release).

File: panic-rtt-target/src/lib.rs (hunk 3 above)
Comment (perlindgren) [reply panic-rtt-target/src/lib.rs:75]: Indeed. I had a compilation error, forcing me to have the extra loop, but I figured it out now. 
TL;DR, I made a stupid mistake :)

I had a `;` after the `critical_section::with(....);` <- here.
Even if Rust is smart enough to figure out that the closure is a ! (and as such would never terminate), Rust type checker goes by concrete types (not inference). thus by removing the `;` the inferred type `!` gets propagated. In this case a "smarter" compiler would have propagate the `effect` of not returning from the closure, but that was not the case.

File: rtt-target/src/rtt.rs (hunk 4 above)
Comment (perlindgren) [reply rtt-target/src/rtt.rs:34]: Ahh now I get it, that's why you had the `_` there as well. If the host uses the symbol table for looking up the location of the RTT datastructure it should not be a problem, if it searches the memory blindly then it might find the magic string in the flash instead of the ram, is that it?

File: rtt-target/src/rtt.rs (hunk 2 above)
Comment (perlindgren) [reply rtt-target/src/rtt.rs:71]: How could this even have worked before?

My assumption here is that its zero initialized. Thus `! ....is_null()` should mean that has been written to. And it works now, so I don't know what I'm missing here.

File: rtt-target/src/rtt.rs (hunk 4 above)
Comment (perlindgren) [reply rtt-target/src/rtt.rs:34]: So the question is how does 'probe` lookup the location (searching through any peekable address, or through the symbol table)?

File: rtt-target/src/rtt.rs (hunk 2 above)
Comment (mvirkkunen) [reply rtt-target/src/rtt.rs:71]: I think this was changed just a couple of days ago and accidentally inverted

File: rtt-target/src/rtt.rs (hunk 4 above)
Comment (mvirkkunen) [reply rtt-target/src/rtt.rs:34]: The symbol table should be and is used whenever available, but there is a fallback to scan the entire memory for the magic string. I think it's a good to keep functional, for example if you need to debug something that you don't have the exact binaries for anymore.

Comment (perlindgren) [reply rtt-target/src/rtt.rs:34]: Ahh got it, yea that makes sense. This fallback should only be active if running without any target file specified right. Not sure what the best thing to do in this case. One could use the string "_EGGER..." and than just write an "S" to the first location. That be similar to your impl. but does not require any pointer offset magic.

File: rtt-target/src/rtt.rs (hunk 2 above)
Comment (perlindgren) [reply rtt-target/src/rtt.rs:71]: Its just a "typo" fix, not sure it rewards a separate PR.

File: rtt-target/src/rtt.rs (hunk 4 above)
Comment (mvirkkunen) [reply rtt-target/src/rtt.rs:34]: That would work as well yes - could maybe even use `ptr::write_volatile` with an array type to ensure the write is actually volatile

Comment (Dirbaio) [reply rtt-target/src/rtt.rs]: Is this change required to make it work on riscv? Or did the previous code also work?

Comment (perlindgren) [reply rtt-target/src/rtt.rs]: I started this to work to add riscv  support to `panic-rtt-target` but ended up here:)

There were some bugs in 0.4.0 of `rtt-target` that needed fixing. When doing so I made some minor changes to leverage the Rust type checker. Still I think one could make the crate less "unsafe" (I just looked a small fraction of the code).

**Comment by perlindgren** [comment]
Are we happy with the suggested PR? 

I did not change versions for `panic-rtt-target`/`rtt-target` yet. Should I do that or do we take that in a separate release round?

**Comment by Dirbaio** [comment]
Could you please clean up the diff? If this PR's goal is "make `panic-rtt-target` work on RISCV" then please do the minimum changes to achieve just that. Other changes should go in separate PRs.

The reason is it's impossible to review otherwise. You mention "there were some bugs in 0.4.0 of rtt-target that needed fixing." Which ones? How are they fixed? I can't tell from the diff alone! The only obvious one I see is the `is_initialized` fix. The other changes seem to be aesthetics: renaming variables, introducing intermediate variables, and changing the code to write `SEGGER RTT` to (as far as I can tell) do the same but in a different way.

Thank you for your understanding.

File: rtt-target/src/rtt.rs (hunk 4 above)
Comment (Dirbaio) [reply rtt-target/src/rtt.rs]: My question was: is *this* change (writing the `SEGGER RTT` string) required for RISCV? As far as I can tell it's doing the same as the previous code, just in a different way.

**Comment by perlindgren** [comment]
It could well be that the only required change to `rtt-target` is the typo fix. I can have a look and make separate PRs.

**Comment by perlindgren** [comment]
Should we consider closing this PR as it was (partly) covered by the merged commit? #36 
https://github.com/probe-rs/rtt-target/commit/a98a8491538d30a8bc3ba72c75c451678c3b1475
